### ISSUE TITLE
java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.pm.PackageManager android.app.Activity.getPackageManager()' on a null object reference

### ISSUE THREAD

**Issue by yushanapp** [g123k/flutter_plugin_device_apps#31]
When calling DeviceApps.getApp the error message is as follows

```
E/MethodChannel#g123k/device_apps(15323): Failed to handle method call
E/MethodChannel#g123k/device_apps(15323): java.lang.NullPointerException: Attempt to invoke virtual method 'android.content.pm.PackageManager android.app.Activity.getPackageManager()' on a null object reference
E/MethodChannel#g123k/device_apps(15323): 	at fr.g123k.deviceapps.DeviceAppsPlugin.getApp(DeviceAppsPlugin.java:165)
E/MethodChannel#g123k/device_apps(15323): 	at fr.g123k.deviceapps.DeviceAppsPlugin.onMethodCall(DeviceAppsPlugin.java:81)
E/MethodChannel#g123k/device_apps(15323): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:231)
E/MethodChannel#g123k/device_apps(15323): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:93)
E/MethodChannel#g123k/device_apps(15323): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:642)
E/MethodChannel#g123k/device_apps(15323): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#g123k/device_apps(15323): 	at android.os.MessageQueue.next(MessageQueue.java:326)
E/MethodChannel#g123k/device_apps(15323): 	at android.os.Looper.loop(Looper.java:190)
E/MethodChannel#g123k/device_apps(15323): 	at android.app.ActivityThread.main(ActivityThread.java:7275)
E/MethodChannel#g123k/device_apps(15323): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#g123k/device_apps(15323): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:575)
E/MethodChannel#g123k/device_apps(15323): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:908)
```

**Comment by g123k** [comment]
Hello,

Could you provide more details on this bug?
Is-it with an app in the background?
Which version of Flutter do you use?

**Comment by ahmadf20** [comment]
Hi, I encountered the same error.  I'm currently using Flutter 1.12.13+hotfix.9.

**Comment by g123k** [comment]
@ahmadf20 Could you provide more info on the context?
Which device? Android version?
After doing what kind of action?

Thanks

**Comment by michael-ottink** [comment]
I am getting the same error when trying to call the DeviceApps from a different isolate https://stackoverflow.com/questions/62820872/how-do-i-correctly-register-a-plugin-with-the-fcm-plugin

**Comment by yushanapp** [comment]
Beginning in Flutter 1.12, the Flutter team adjusted the implementation code of the Android plug-in. After 1.12, Android began to use the new plug-in API. The old PluginRegistry.Registrar based on it will not be immediately deprecated, but the official recommendation is to migrate to the new API based on FlutterPlugin. In addition, the new version of the official recommended plug-in directly uses Androidx support, and the officially provided plug-in has also been fully upgraded to Androidx.

Compared with the old API, the advantage of the new API is that it provides a more decoupled use method for the life cycle that the plug-in depends on. For example, when PluginRegistry.Registrar.activity() was used before, if Flutter has not been added to Activity may return null at the same time, and the plug-in does not know when it is loaded by the engine, and these problems have been optimized on the new API

For me, after modifying the plug-in registration method, it can be used normally. The following is my modification

```

public class DeviceAppsPlugin implements MethodCallHandler, FlutterPlugin, ActivityAware, PluginRegistry.ViewDestroyListener {

    private FlutterPluginBinding pluginBinding;
    private Context applicationContext;
    private Activity activity;
    private BinaryMessenger messenger;

    /**
     * Plugin registration.
     */
    // public static void registerWith(Registrar registrar) {
    //     final MethodChannel channel = new MethodChannel(registrar.messenger(), "g123k/device_apps");
    //     DeviceAppsPlugin plugin = new DeviceAppsPlugin(registrar.activity());
    //     registrar.addViewDestroyListener(plugin);
    //     channel.setMethodCallHandler(plugin);
    // }

    private final int SYSTEM_APP_MASK = ApplicationInfo.FLAG_SYSTEM | ApplicationInfo.FLAG_UPDATED_SYSTEM_APP;

    private  AsyncWork asyncWork;

    private void initPlugin(Context applicationContext, Activity activity, BinaryMessenger messenger) {
        this.applicationContext = applicationContext;
        this.activity = activity;
        this.messenger = messenger;
        this.asyncWork = new AsyncWork();

        final MethodChannel channel = new MethodChannel(messenger, "g123k/device_apps");
        channel.setMethodCallHandler(this);
    }


    @Override
  public void onAttachedToEngine(FlutterPluginBinding flutterPluginBinding) {
    Log.e("onAttachedToEngine", "onAttachedToEngine");
    pluginBinding = flutterPluginBinding;
  }

  @Override
  public void onDetachedFromEngine(FlutterPluginBinding binding) {
    Log.e("onDetachedFromEngine", "onDetachedFromEngine");
    pluginBinding = null;
  }


  ///activity 
  @Override
  public void onAttachedToActivity(ActivityPluginBinding activityPluginBinding) {
    Log.e("onAttachedToActivity", "onAttachedToActivity");
    initPlugin(pluginBinding.getApplicationContext(), activityPluginBinding.getActivity(), pluginBinding.getBinaryMessenger());

  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
    Log.e("onDetachedFromActivityForConfigChanges", "onDetachedFromActivityForConfigChanges");
    activity = null;
    asyncWork.stop();
  }

  @Override
  public void onReattachedToActivityForConfigChanges(ActivityPluginBinding activityPluginBinding) {
    Log.e("onReattachedToActivityForConfigChanges", "onReattachedToActivityForConfigChanges");
    initPlugin(pluginBinding.getApplicationContext(), activityPluginBinding.getActivity(), pluginBinding.getBinaryMessenger());
  }

  @Override
  public void onDetachedFromActivity() {
    Log.e("onDetachedFromActivity", "onDetachedFromActivity");
    activity = null;
    asyncWork.stop();
  }
  ......
}

```

**Comment by g123k** [comment]
The problem should be fixed with 1.1.0.
Please let me know

**Comment by michael-ottink** [comment]
I am still getting the following error using 1.1.0

```
E/flutter (24866): [ERROR:flutter/lib/ui/ui_dart_state.cc(157)] Unhandled Exception: MissingPluginException(No implementation found for method openApp on channel g123k/device_apps)
E/flutter (24866): #0      MethodChannel._invokeMethod 
package:flutter/…/services/platform_channel.dart:154
E/flutter (24866): <asynchronous suspension>
E/flutter (24866): #1      MethodChannel.invokeMethod 
package:flutter/…/services/platform_channel.dart:329
E/flutter (24866): #2      DeviceApps.openApp 
package:device_apps/device_apps.dart:81
E/flutter (24866): #3      SystemOverlayService.systemOverlayOnClickListner 
package:app_name/…/system_overlay/system_overlay_service.dart:42
E/flutter (24866): #4      callbackDispatcher.<anonymous closure> 
package:system_alert_window/system_alert_window.dart:136
E/flutter (24866): #5      MethodChannel._handleAsMethodCall 
package:flutter/…/services/platform_channel.dart:409
E/flutter (24866): #6      MethodChannel.setMethodCallHandler.<anonymous closure> 
package:flutter/…/services/platform_channel.dart:377
E/flutter (24866): #7      _DefaultBinaryMessenger.handlePlatformMessage 
package:flutter/…/services/binding.dart:199
E/flutter (24866): #8      _invoke3.<anonymous closure>  (dart:ui/hooks.dart:290:15)
E/flutter (24866): #9      _rootRun  (dart:async/zone.dart:1184:13)
```
